### PR TITLE
Introducing StartableByRPC and SchedulableFlow annotations, needed by…

### PR DESCRIPTION
**Pull request by shamsasari**
… flows started via RPC and schedulable flows respectively.

CordaPluginRegistry.requiredFlows is no longer needed as a result.

### CODE REVIEW

**Comment by shamsasari** [comment]
`FlowLogicRefFactoryImpl.flowWhitelist` has been removed as it's not needed, at least the keys of the map. I couldn't see why would need to whitelist the arg types so they're gone as well. Let me if they are needed and I've missed something .

**Comment by shamsasari** [comment]
And also if you think of better names for the annotations.

File: docs/source/changelog.rst [hunk 1]
```diff
@@ -8,8 +8,11 @@ UNRELEASED
 ----------
 
 * API changes:
+    * ``CordaPluginRegistry.requiredFlows`` is no longer needed. Instead annotate any flows you wish to start via RPC with
+      ``@StartableByRPC`` and any scheduled flows with ``@SchedulableFlow``.
+
     * Initiating flows (i.e. those which initiate flows in a counterparty) are now required to be annotated with
```
Comment (mikehearn) [inline docs/source/changelog.rst]: This phrasing is still very confusing to me. All flows "initiate flows in a counterparty" in some sense don't they.

File: docs/source/flow-state-machines.rst [hunk 2]
```diff
@@ -206,9 +206,9 @@ how to register handlers with the messaging system (see ":doc:`messaging`") and
 when messages arrive. It provides the send/receive/sendAndReceive calls that let the code request network
 interaction and it will save/restore serialised versions of the fiber at the right times.
 
-Flows can be invoked in several ways. For instance, they can be triggered by scheduled events,
-see ":doc:`event-scheduling`" to learn more about this. Or they can be triggered directly via the Java-level node RPC
-APIs from your app code.
+Flows can be invoked in several ways. For instance, they can be triggered by scheduled events (in which case they need to
+be annotated with ``@SchedulableFlow``), see ":doc:`event-scheduling`" to learn more about this. Or they can be triggered
+directly via the node's RPC API from your app code (in which case they need to be annotated with `StartableByRPC`).
```
Comment (mikehearn) [inline docs/source/flow-state-machines.rst]: This phrasing makes it sound mutually exclusive, it can _either_ be scheduled _or_ via RPC. But it's reasonable for a flow to be both (in fact, arguably that's the common case). Let's rephrase this to make it clear you can have both.

File: docs/source/flow-state-machines.rst [hunk 3]
```diff
@@ -399,15 +399,50 @@ This code is longer but no more complicated. Here are some things to pay attenti
...
+do not interfere with each other.
+
+To achieve this the flow framework initiates a new flow session each time a flow starts communicating with a ``Party``
+for the first time. A session is simply a pair of IDs, one for each side, to allow the node to route received messages to
+the correct flow. If the other side accepts the session request then subsequent sends and receives to that same ``Party``
+will use the same session. A session ends when either flow ends, whether as expected or pre-maturely. If a flow ends
+pre-maturely then the other side will be notified of that and they will also end, as the whole point of flows is a known
+sequence of message transfers. Flows end pre-maturely due to exceptions, and as described above, if that exception is
+``FlowException`` or a sub-type then it will propagate to the other side. Any other exception will not propagate but will
+be logged.
```
Comment (mikehearn) [inline docs/source/flow-state-machines.rst]: Both exception types are logged, no?

File: docs/source/flow-state-machines.rst [hunk 4]
```diff
@@ -399,15 +399,50 @@ This code is longer but no more complicated. Here are some things to pay attenti
...
+
+Taking a step back, we mentioned that the other side has to accept the session request for there to be a communication
+channel. A node accepts a session request if it has registered the flow type (the fully-qualified class name) that is
+making the request - each session initiation includes the initiating flow type. The registration is done by a CorDapp
+which has made available the particular flow communication, using ``PluginServiceHub.registerServiceFlow``. This method
+specifies a flow factory for generating the counter-flow to any given initiating flow. If this registration doesn't exist
+then no further communication takes place and the initiating flow ends with an exception. The initiating flow has to be
+annotated with ``InitiatingFlow``.
+
+To recap, once a session has been estabished with a ``Party`` any subsequent sends or receives will be done on that session.
```
Comment (mikehearn) [inline docs/source/flow-state-machines.rst]: I find this paragraph a bit confusing still. You've introduced the term "counter flow" without defining it.

I think the underlying issue here is that it's not really obvious how or why you'd want to start a new session. In theory, every subflow could be an inline flow with shared parent sessions, right? Then the top level flows would always be because they were started by RPC/scheduler/peer and there'd be no need to think about any of this. It'd mean adjusting some parts of the existing flows to understand that they're about to be interrogated for tx resolution or attachment fetches or whatever, but it's not infeasible, right?

I can't escape the feeling that this session stuff is just more conceptually complex than necessary. I realise it was written to fit how the existing code worked, but I suspect the original code was more complex than necessary too.

What's described in this paragraph is likely to either be an advanced use case, or one we can remove entirely. I think that's why it feels confusing to me.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 5]
```diff
@@ -132,10 +127,10 @@ abstract class AbstractNode(open val configuration: NodeConfiguration,
         override val myInfo: NodeInfo get() = info
         override val schemaService: SchemaService get() = schemas
         override val transactionVerifierService: TransactionVerifierService get() = txVerifierService
+        override val rpcFlows: List<Class<out FlowLogic<*>>> get() = this@AbstractNode.rpcFlows
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Is this necessary only for the glob completion in the shell? Where do we actually need access to an entire list of every flow in the system? The rest of the time the class name is always specified completely, right?

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt [hunk 6]
```diff
@@ -115,26 +115,26 @@ class CordaRPCOpsImpl(
...
     }
 
-    // TODO: Check that this flow is annotated as being intended for RPC invocation
     override fun <T : Any> startFlowDynamic(logicType: Class<out FlowLogic<T>>, vararg args: Any?): FlowHandle<T> {
+        val stateMachine = startFlow(logicType, args)
+        return FlowHandleImpl(id = stateMachine.id, returnValue = stateMachine.resultFuture)
+    }
+
+    private fun <T : Any> startFlow(logicType: Class<out FlowLogic<T>>, args: Array<out Any?>): FlowStateMachineImpl<T> {
+        require(logicType.isAnnotationPresent(StartableByRPC::class.java)) { "${logicType.name} is not permissioned for RPC" }
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: It's not so much a security issue, actually. Flows already have permissions associated with them. It's more like a safety issue: it doesn't make conceptual sense to start a flow that's intended to be only a subflow via RPC or the shell, so if they're explicitly marked, we can check you aren't making a conceptual mistake earlier (and filter out flows-that-are-only-subflows from UIs/completion/etc).

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt [hunk 7]
```diff
@@ -315,6 +313,29 @@ abstract class AbstractNode(open val configuration: NodeConfiguration,
...
+        log.info("Scanning plugins in $pluginsDir ...")
+        if (!pluginsDir.exists()) return emptySet()
+
+        val pluginJars = pluginsDir.list {
+            it.filter { it.isRegularFile() && it.toString().endsWith(".jar") }.collect(toList())
+        }
+
+        fun Reflections.scanFlows() = getSubTypesOf(FlowLogic::class.java).filter { isPublic(it.modifiers) && !isAbstract(it.modifiers) }
+
+        val flows = pluginJars.flatMapTo(HashSet()) { jar ->
```
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Does this make startup a lot slower? I guess most plugins are small, so, no?

Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: I know I mentioned reflections, sorry about that, but it has an awkward license and the maintainer has refused to relicense it under something more conventional.

Could we take a look at https://github.com/lukehutch/fast-classpath-scanner instead?

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt (hunk 5 above)
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Ah right, we expose the full list via RPC for diagnostics and UI purposes I guess. OK, it's fine.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt (hunk 7 above)
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: Startup time seems to be affected very little.

File: node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt (hunk 6 above)
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: So should I remove this check or just reword the message?

Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: If the check is not there then no-one will bother using the annotation.

Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/CordaRPCOpsImpl.kt]: I didn't mean remove it, sorry, I meant change the message.

File: node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt (hunk 7 above)
Comment (mikehearn) [inline node/src/main/kotlin/net/corda/node/internal/AbstractNode.kt]: OK, that's good to hear.

File: docs/source/flow-state-machines.rst (hunk 4 above)
Comment (shamsasari) [inline docs/source/flow-state-machines.rst]: OK, I'll remove this paragraph and think about removing the need for `InitiatingFlow` (at least as a required annotation). Currently these are the sub-flows that need to be annotated with this:
- NotaryClient.Flow
- BroadcastTransactionFlow
- FetchTransactionsFlow
- FetchAttachmentsFlow
- FixQueryFlow
- FixSignFlow
- SimmFlow.Requester

**Comment by shamsasari** [comment]
Comments addressed.

File: docs/source/flow-state-machines.rst (hunk 4 above)
Comment (mikehearn) [inline docs/source/flow-state-machines.rst]: OK. It may be that I just need to get more comfortable with the idea.

Of the above flows, how many need to be context-free/top-level *and* have an already existing session in place when they're used? NotaryClient.Flow is always initiating, but a flow wouldn't have a session in place normally, I think? Unless a single flow notarises multiple transactions, which I suppose, is possible. Hmm.

Perhaps this notion is required, and we just need to find a way to make it clearer in the docs.

Comment (shamsasari) [inline docs/source/flow-state-machines.rst]: In the PR following this, what I'll do is make `InitiatingFlow` optional. It'll only be required in two cases:
1. If you upgrade your flow then you'll need to use to specify the `version`. This will save us from having a separate `FlowVersion` annotation.
2. Needed for sub-flows that must start a new season, like the notary client.

I'll also have a look and see if any of the above flows can have their initiating annotation removed. As you say, some like the notary client will need it.

Let me know if the current docs still need more work.

Comment (mikehearn) [inline docs/source/flow-state-machines.rst]: When inlining a subflow, the ability to know the version of the other side is still useful. I realise that it's not supported at present (would require sending a "starting subflow" message  I guess?), but otherwise I'm not sure how we can evolve subflows designed to be inlined like CollectSignatures.

File: node/build.gradle [hunk 8]
```diff
@@ -150,6 +150,9 @@ dependencies {
     // Requery: object mapper for Kotlin
     compile "io.requery:requery-kotlin:$requery_version"
 
+    // Reflections: classpath scanning
```
Comment (mikehearn) [inline node/build.gradle]: This isn't Reflections anymore.

File: docs/source/flow-state-machines.rst (hunk 4 above)
Comment (shamsasari) [inline docs/source/flow-state-machines.rst]: The obvious thing to do is bump the parent flow's version number.

File: core/src/main/kotlin/net/corda/core/flows/SchedulableFlow.kt [hunk 9]
```diff
@@ -0,0 +1,14 @@
...
+
+/**
+ * Any [FlowLogic] which is schedulable and is designed to be invoked by a [net.corda.core.contracts.SchedulableState]
+ * must have this annotation. If it's missing [FlowLogicRefFactory.create] will throw an exception when it comes time
+ * to schedule the next activity in [net.corda.core.contracts.SchedulableState.nextScheduledActivity].
+ */
+@Target(CLASS)
+@Inherited
+@MustBeDocumented
+annotation class SchedulableFlow
```
Comment (rick-r3) [inline core/src/main/kotlin/net/corda/core/flows/SchedulableFlow.kt]: Should this be `StartableByScheduler` to be more consistent with `StartableByRPC` ?  Or the other way around and `StartableByRPC` is `RemoteStartableFlow` ?

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowLogicRefFactoryImpl.kt [hunk 10]
```diff
@@ -30,58 +26,23 @@ data class FlowLogicRefImpl internal constructor(val flowLogicClassName: String,
...
-    /**
-     * Create a [FlowLogicRef] for the Kotlin primary constructor of a named [FlowLogic]
-     */
-    fun createKotlin(flowLogicClassName: String, args: Map<String, Any?>, attachments: List<SecureHash> = emptyList()): FlowLogicRef {
-        val context = AppContext(attachments)
-        validateFlowClassName(flowLogicClassName, context)
-        for (arg in args.values.filterNotNull()) {
-            validateArgClassName(flowLogicClassName, arg.javaClass.name, context)
+object FlowLogicRefFactoryImpl : SingletonSerializeAsToken(), FlowLogicRefFactory {
+    override fun create(flowClass: Class<out FlowLogic<*>>, vararg args: Any?): FlowLogicRef {
```
Comment (rick-r3) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowLogicRefFactoryImpl.kt]: I think we should add a TODO somewhere in this class to suggest that we will at some point check whether there is permission, beyond the annotation, to start the flows.  For example, as a security measure we might want the ability for the node admin to blacklist a flow such that it moves immediately to the "Flow Hospital" in response to a potential malicious use or buggy update to an app etc.

How many of those TODOs on this class remain valid?  e.g. the one about how to populate the whitelist

File: core/src/main/kotlin/net/corda/core/flows/SchedulableFlow.kt (hunk 9 above)
Comment (shamsasari) [inline core/src/main/kotlin/net/corda/core/flows/SchedulableFlow.kt]: I'll add a TODO to `StartableByRPC` as I like `SchedulableFlow`

File: node/src/main/kotlin/net/corda/node/services/statemachine/FlowLogicRefFactoryImpl.kt (hunk 10 above)
Comment (shamsasari) [inline node/src/main/kotlin/net/corda/node/services/statemachine/FlowLogicRefFactoryImpl.kt]: Done
